README.md
Thoom::RestUp
=============
[![Gem Version](https://badge.fury.io/rb/restup.svg)](https://badge.fury.io/rb/restup) [![Build Status](https://travis-ci.org/thoom/restup.svg?branch=master)](https://travis-ci.org/thoom/restup)

After many years in beta, RestClient is now RestUp version 1. 
It's mainly the same client as before, but some breaking changes have been made based on 
feedback and experience.

RestUp works out of the box with APIs that use Basic Authentication (though this is not required).
To use other forms of authentication, custom headers can either be passed with each request
or stored in the config file as described below.

If a YAML configuration file exists (by default `.restup.yml`), the client will pull in defaults and provide several shortcut methods that can simplify using a REST-based API.

If the API uses form encoded input, you can define your post in JSON format. The client
will encode it automatically.

Installation
------------

#### Gem
For convenience, the executable and class are available as a gem on RubyGems.

    gem install restup

#### Docker
The client is also available as a Docker image.

To install:

    docker pull thoom/restup

To run:

    docker run --rm -v $PWD:/restup thoom/restup

A sample shell script `restup`:

    #!/bin/bash
    docker run --rm -v $PWD:/restup thoom/restup "@"

Console
-------

    Usage: restclient [options] ENDPOINT
            --concise                    Simplified response view
            --content-disposition        For responses with a filename in the Content Disposition, save the response using that filename
            --form                       Converts JSON-formatted input and encode it as x-www-form-urlencoded
            --response-only              Only outputs the response body
            --response-code-only         Only outputs the response code
            --success-only               Only outputs whether or not the request was successful
            --cert CERTFILE              Imports cert for Client-Authentication endpoints
        -c, --config FILE                Config file to use. Defaults to .restup.yml
        -e, --env ENVIRONMENT            Sets YAML environment for the request
        -h, --header HEADER              Sets arbitrary header passed in format "HEADER: VALUE"
        -j, --json [c|a]                 Sets the Content-Type and/or Accept Headers to use JSON mime types (i.e. -ja)
        -m, --method                     The HTTP method to use (defaults to GET)
        -o, --output FILE                Save output to file passed
        -p, --password PASSWORD          Password for Basic Authentication
        -u, --username USERNAME          Username for Basic Authentication
        -x, --xml [c|a]                  Sets the Content-Type and/or Accept Headers to use XML mime types (i.e. -xc)
            --verbose                    Expanded response view
            --version                    Shows client version
            --help [details]             Shows this help dialog

YAML config
-----------

If a file is not passed in with the `-c, --config` flag, then it will use the default `.restup.yml`.
The client uses two different methods to find the YAML configuration file.
It will first look in the current directory. 
If it is not present, it will then look in the current user's home directory.

This makes it possible to use restup to connect to different APIs simply by changing
folders.

    KEY          DESC
    ----         -----
    user:       The username. Default: blank, so disable Basic Authentication
    pass:       The password. Default: blank, so disable Basic Authentication

    url:        The base REST url
    json:       The default JSON MIME type. Default: "application/json"
    xml:        The default XML MIME type.  Default: "application/xml"

    colors:     Hash of default color values
      success:  Color to highlight successful messages.  Default: :green
      warning:  Color to highlight warning messages.     Default: :yellow
      info:     Color to highlight info messages.        Default: :yellow
      error:    Color to highlight error messages.       Default :red
      
    flags:      Default command line options
      display:  What to display by default. 
                Values: concise, response_only, response_code_only, succcess_only, verbose
                Default: response_only

    headers:    Hash of default headers. Useful for custom headers or headers used in every request.
    timeout:    The number of seconds to wait for a response before timing out. Default: 300
    tls_verify: Whether or not to verify TLS (SSL) certificates.  Values: true, false.  Default: true

    xmethods:   Array of nonstandard methods that are accepted by the API. To use these methods the
                API must support X-HTTP-Method-Override.

Examples
--------

### GET Request

The YAML config:

    url: http://example.com/api
    user: myname
    pass: P@ssWord

The command line:

    restup -j /hello/world 

To use without the config:

    restclient -u myname -p P@ssWord -j http://example.com/api/hello/world

Submits a GET request to `http://example/api/hello/world` with Basic Auth header using the
user and pass values in the config.

It would return JSON values. If successful, the JSON would be parsed and highlighted in __colors['success']__. If
the an error was returned (an HTTP response code >= 400), the body would be in __colors['error']__.

### POST Request

The YAML config:

    url: http://example.com/api
    user: myname
    pass: P@ssWord
    headers:
      X-Custom-Id: abc12345

The command line:

    restclient -m post -j /hello/world < salutation.json

OR

    cat salutation.json | restclient -m post /hello/world

Submits a POST request to `http://example/api/hello/world` with Basic Auth header using the
user and pass values in the config. It imports the salutation.json and passes it to the API as application/json
content type. It would also set the `X-Custom-Id` header with every request.

It would return JSON values. If successful, the JSON would be parsed and highlighted in __colors['success']__. If
the an error was returned (an HTTP response code >= 400), the body would be in __colors['error']__.


Migration From RestClient
-------------------------

To migrate:

1. Rename your `.restclient.yml` file to `.restup.yml`.
2. The CLI format changed from `restclient METHOD ENDPOINT [options]` to `restup [options] ENDPOINT`.
3. The `-c` option is no longer available. You must use `--cert` instead.
4. The `-m` option was created for specifying methods. So `restup -m POST ENDPOINT` instead of `restclient POST ENDPOINT`.
5. Add `flags: { display: verbose }` to the config file return to the previous API output.

License
-------
[MIT](LICENSE)

Version History
---------------
[Changelog](CHANGELOG.md)


CHANGELOG.md
### Version Changelog
1.1.3 Updated to use [Paint v. 2.0](https://github.com/janlelis/paint) gem.

1.0.1 Fixes issue [#33](https://github.com/thoom/restclient/issues/33).

1.0.0 RestClient => RestUp. _Finally_ out of beta.
 - Fixes Github issue [#30](https://github.com/thoom/restclient/issues/30) by renaming RestClient to RestUp.
 - Fixes Github issue [#28](https://github.com/thoom/restclient/issues/28). `--form` converts JSON to form data only. 
   To use form data, manually set the header using the `-h` option or set the header in the configuration.
 - Fixes Github issues [#24](https://github.com/thoom/restclient/issues/24) and [#25](https://github.com/thoom/restclient/issues/25). Use `--verbose` or set `flags: { display: verbose }` in config.
 - Fixes Github issue [#31](https://github.com/thoom/restclient/issues/31). Use `-c` or `--config` to pass in another config file.
 - Fixes Github issue [#20](https://github.com/thoom/restclient/issues/20).
 - Fixes Github issue [#11](https://github.com/thoom/restclient/issues/11).


0.14.3 Fixes Github issues [#21](https://github.com/thoom/restclient/issues/21) and [#26](https://github.com/thoom/restclient/issues/26). 
You can use YAML now instead of JSON for JSON requests. Since JSON is a subset of YAML, this is not a breaking change.

0.14.2 Fixes Github issue [#17](https://github.com/thoom/restclient/issues/17). It also simplifies how files or data are passed into the client.
Breaking change because flags -f and -d have been removed. Instead just pass in with STDIN instead.

0.13.0 Fixes Github issue [#13](https://github.com/thoom/restclient/issues/13). 
Breaking change since it no longer creates a config file if one doesn't exist.

0.12.1: Fixes Github issues [#12](https://github.com/thoom/restclient/issues/12) and [#19](https://github.com/thoom/restclient/issues/19). 
If using the library class instead of the CLI, you can pass in a `Thoom::HashConfig` object,
instead of requiring a YAML file. CLI no longer generates a deprecated warning in the Ruby 2.3.

0.11.3: Adds a simple response time to the output.

0.11.2: Fixes Github issue [#10](https://github.com/thoom/restclient/issues/10).

0.11.1: Fixes Github issues [#2](https://github.com/thoom/restclient/issues/2), [#5](https://github.com/thoom/restclient/issues/5), [#6](https://github.com/thoom/restclient/issues/6), [#7](https://github.com/thoom/restclient/issues/7), and [#8](https://github.com/thoom/restclient/issues/8). If a configuration file is not found,
there is now a simple wizard that will ask a few questions and then generate the config file.

0.10.1: Fixes Github issues [#1](https://github.com/thoom/restclient/issues/1) and [#4](https://github.com/thoom/restclient/issues/4). `PUT`, `PATCH`, and `POST` all set a default `JSON` content type. Added `--response-only` flag.

0.10.0: Fixed a bug missed where headers in the YAML file were not parsed. Changed the YAML :headers: to use a hash instead of array.

0.9.3: Rewrote console output mechanism. Fixed a few bugs related to YAML configuration. Added `--response-code-only` and `--success-only` flags.

0.9.0: Significant internal refactoring to the client. Also changed console coloring dependency to the `Paint` gem instead of `Colored` gem.

0.8.5: Added option to save response to the filename in the content-disposition if that header is passed.

0.8.4: Added option to save response to a file by passing in the -o {file} arg. For binary files, response is never output in summary.

0.7.5: Added option to change the request timeout. Added option to disable TLS cert validation (Useful with self-signed certs).

0.7.0: Removed dependency on Nokogiri for XML parsing.


test/config_test.rb
require 'byebug' if ENV['DEBUG']
require 'minitest/autorun'
require_relative '../lib/config.rb'

# Test the configuration system
class TestConfig < Minitest::Test
  def setup
    hash = {
      url: 'http://github.com/thoom/restup',
      mime_json: 'application/json',
      second_env: { url: 'http://github.com/thoom' }
    }
    @config = Thoom::HashConfig.new hash
  end

  def test_default_env
    @config.env = :default
    assert_equal 'http://github.com/thoom/restup', @config.get(:url)
    assert_equal 'application/json', @config.get(:mime_json)
  end

  def test_second_env
    @config.env = :second_env
    assert_equal 'http://github.com/thoom', @config.get(:url)
    assert_equal 'application/json', @config.get(:mime_json)
  end

  def test_default_value
    assert_equal 'default_val', @config.get(:missing_key, 'default_val')
  end

  def test_add_value
    @config.set(:new_key, 'new value')
    assert_equal 'new value', @config.get(:new_key)

    @config.env = :second_env
    assert_equal 'new value', @config.get(:new_key)
  end

  def test_missing_raises
    assert_raises(Thoom::ConfigError) { @config.get(:missing_raises) }
  end
end


test/rest_up_test.rb
require 'base64'

require 'byebug' if ENV['DEBUG']
require 'minitest/autorun'
require 'webmock/minitest'
require_relative '../lib/rest_up'

CLIENT_CERT = '-----BEGIN CERTIFICATE-----
MIIC/jCCAeYCAQIwDQYJKoZIhvcNAQELBQAwRTELMAkGA1UEBhMCQVUxEzARBgNV
BAgMClNvbWUtU3RhdGUxITAfBgNVBAoMGEludGVybmV0IFdpZGdpdHMgUHR5IEx0
ZDAeFw0xNzAzMjAwMDM0NDVaFw0xNzA0MTkwMDM0NDVaMEUxCzAJBgNVBAYTAkFV
MRMwEQYDVQQIDApTb21lLVN0YXRlMSEwHwYDVQQKDBhJbnRlcm5ldCBXaWRnaXRz
IFB0eSBMdGQwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCuyAWJjsSv
HwKAy0Zl1O6rpGfWk9gsWBbpPVHPvvf2spXd+LhrPfZ0SH8/Lxst82DMGTXTLtLC
COsQGnm6bMrFEwlHpVvGmEo24sa5HyRhGWhsbnP0DML9CtLrdrDQaluZxEOBBWAl
AYkk1OtJIlyXaKywFTwtt+KdU6R1R29Elnn3Mzdy+9s2pAuyGfHOFQo+zmaUWjq3
bzumfXkggSXTH46KCVk5vThnA48E3lxudOLEkuH5OM27UthbfU7HxwlmHm8lXeE1
EyMc3RL4HC2YECRLj3V78VKc+tBQfLS/TkFCUALbIYYg+zZ/gaOW3Rhf+sGjXyot
OQkHmjb0GVDxAgMBAAEwDQYJKoZIhvcNAQELBQADggEBAM8/mnngdWaAi1xbywuL
D/ifvbvRbeE/YGmR968lVWC5UxnOuHHGPdavbLr/cMVyAeuNFCz+TcmSfIkLQnE3
A2iRrEDoTLPXlXy0EOgMf+4RtOzG7v74CRPwPVPBe99T7PIWVQZJMQgi7aM7ypxK
VP/eVhO73DQT2fBqegdXCujBHvZKzOH3FC7UhyZkKiMnzFgVyUdIA7EwKwSwBj1j
dNjfpsSPBdeiiRI4xBIPw9Pwt0bSpE9SRe+iqfc+mrdpFbQFjLLeGfKGbzPw2IFB
WRvWtj6hcU0X0y9WrdRUUKmAAmfAn6NV8dSys6xL0N3WNspEayl+HNglUm243fgg
P1c=
-----END CERTIFICATE-----
-----BEGIN RSA PRIVATE KEY-----
MIIEowIBAAKCAQEArsgFiY7Erx8CgMtGZdTuq6Rn1pPYLFgW6T1Rz7739rKV3fi4
az32dEh/Py8bLfNgzBk10y7SwgjrEBp5umzKxRMJR6VbxphKNuLGuR8kYRlobG5z
9AzC/QrS63aw0GpbmcRDgQVgJQGJJNTrSSJcl2issBU8LbfinVOkdUdvRJZ59zM3
cvvbNqQLshnxzhUKPs5mlFo6t287pn15IIEl0x+OiglZOb04ZwOPBN5cbnTixJLh
+TjNu1LYW31Ox8cJZh5vJV3hNRMjHN0S+BwtmBAkS491e/FSnPrQUHy0v05BQlAC
2yGGIPs2f4Gjlt0YX/rBo18qLTkJB5o29BlQ8QIDAQABAoIBAEzHxdb2BuU/ZCQc
KGLXtbllUdBrxk1ErVvs+iGfLOMM9NU8TBegcsIGjw+8q931ypIeiQyqxx9CN7Et
UEhbPoBwqmT0+Wo2wHN4EACG875MIRlxNDMD4Cy1Mo8J5wbHklHsbBMjv2bCU/MA
LhREqhRBgYsg6PzUiRLACcETjoe8MJABTYQTBhJ16Pk046zP+DXdzPu36FnLDVhy
j9JU0nlu6QDTkUq1Zv6J5e1z57vkVOe8oJuvyomyG/2ziCUBoibIU0kyLoOMTTVi
shlmrQUAhmJPugzYop2a49Qj/LRIEc1dCIPTfZ4nRGq0f5AzLU9tUt54F925tQfJ
D1psAaECgYEA5NGKEKDAx/oATBvVIWnzgsr6iq/VJfUpa1SFvB/kjjtDUA861qsS
niQLI9557HlF/HaSgwG07lUUuCfrmt3Og42Dt4apWk+/WLOy74Xb/lo+aUWz5qFs
UOCCeB1YOzdCZ8kZUhHpQonOvpih71d6iSw1g/0L2wu69gaJgZoETY0CgYEAw4sw
xr10TXdTrBqHfvqWHJKFx6QZM/jtmGBWhjzXGB2VVedNLxzOsFmMTkZMWOlbNcxg
3RZ9wMMfXMHqiPX+peObv4UwoD2HSUjH54yQZ3oqDRagfoplkc1Hw4jYUPtMFZsn
4KuprV9lNj5l+HXr+ve1W2wA14kNXuABDGQivfUCgYA8bEDDJ1AA/rl5X5gmXK4b
CbKjUM8+WMD5QLaX+OwHywp5Z6wn58Dg1a/DZwpXMacThdQmmBrJHNp9zrzehlf0
UThJDFxIJurmZ32G1phDUF1Ou0NIUbQin0aUpVsZN/xnH6l6DJTGJ1Ha7r9ETptj
AbOYCQhKbYyPTyacfKlKOQKBgQCLBS0zTkQeQwSwqdfE9eC9BYqo6cilQ4efunYp
T53YXkfqX9xm70Me6zst9xqWZ6lZ8Si4ZiIXZmGor5DPuJxHUi9LlSSB99xzxJOi
0jNj7d1xmrGV5UzawKgRovuvb0mjXsCWIVUrllO9odUbNLMFpRBBo+JhWeWwmu8D
4Tk19QKBgFjx//9+Mz+XlMwHA8g+jVRba0PSPRp2W4MyjAzHWZeJ/1JyA9d6YLn+
cmqFENP4evjroplLtt+SQJk5xaiEyW+BsnZNkEzRiHpqhm4OCTbotyQyxTbR9g4P
ASSPbJlaDNnYBc/lxHPOcpHsXJF0R7DZlOzwbY8dBMt7pKoWtEev
-----END RSA PRIVATE KEY-----'.freeze

# Test the restup class
class TestRestUp < Minitest::Test
  def test_url
    restup = Thoom::RestUp.new(Thoom::HashConfig.new(url: 'http://ex.io'))

    # Standard case
    restup.endpoint = '/test/'
    assert_equal 'http://ex.io/test/', restup.url

    # If the endpoint starts with http, don't append the config's URL
    restup.endpoint = 'http://someotherdomain.com/test/'
    assert_equal 'http://someotherdomain.com/test/', restup.url
  end

  def test_invalid_url
    restup = Thoom::RestUp.new(Thoom::HashConfig.new(url: 'invalid_url'))

    restup.endpoint = '/foo'
    error = assert_raises(Thoom::RestUpError) { restup.request }
    assert_equal 'Invalid URL', error.message
  end

  def test_basic_auth
    config = Thoom::HashConfig.new(
      user: 'joe', pass: 'abc1234', url: 'http://ex.io'
    )

    restup = Thoom::RestUp.new(config)
    restup.method = 'get'
    restup.endpoint = '/foo'

    encoded = Base64.urlsafe_encode64('joe:abc1234')

    request = restup.request
    assert_equal "Basic #{encoded}", request.each_header.to_h['authorization']
  end

  def test_invalid_method
    restup = Thoom::RestUp.new(Thoom::HashConfig.new(url: 'http://ex.io'))

    error = assert_raises(Thoom::RestUpError) { restup.method = 'foobar' }
    assert_equal 'Invalid Method', error.message
  end

  def test_xmethod
    config = Thoom::HashConfig.new(url: 'http://ex.io', xmethods: ['FooBar'])
    restup = Thoom::RestUp.new(config)
    restup.method = 'foobar'

    # If the method is an allowed xmethod,
    # then change it to post and add a method override header
    assert_equal 'post', restup.method
    assert_equal 'FOOBAR', restup.headers['x-http-method-override']

    # If the method is not an allowed xmethod, it should raise an error
    error = assert_raises(Thoom::RestUpError) { restup.method = 'fezbaz' }
    assert_equal 'Invalid Method', error.message
  end

  def test_xmethod_error
    # xmethods should always be an array
    config = Thoom::HashConfig.new(url: 'http://ex.io', xmethods: 'FooBar')
    restup = Thoom::RestUp.new(config)

    error = assert_raises(Thoom::RestUpError) { restup.method = 'foobar' }
    assert_equal 'Invalid xmethods configuration', error.message
  end

  def test_client_cert
    config = Thoom::HashConfig.new(url: 'https://ex.io', cert: CLIENT_CERT)
    restup = Thoom::RestUp.new(config)

    restup.method = 'get'
    restup.endpoint = '/api'

    http = restup.http
    assert http.cert
    assert http.key
  end

  def test_client_cert_invalid
    config = Thoom::HashConfig.new(url: 'https://ex.io', cert: 'bad')
    restup = Thoom::RestUp.new(config)

    restup.method = 'get'
    restup.endpoint = '/api'

    error = assert_raises(Thoom::RestUpError) { restup.http }
    assert_equal 'Invalid client certificate', error.message
  end

  def test_tls
    config = Thoom::HashConfig.new(url: 'https://ex.io')
    restup = Thoom::RestUp.new(config)

    restup.method = 'get'
    restup.endpoint = '/api'

    # Default to verifying SSL
    http = restup.http
    assert_equal true, http.use_ssl?
    assert_equal OpenSSL::SSL::VERIFY_PEER, http.verify_mode
  end

  def test_tls_verify_off
    config = Thoom::HashConfig.new(url: 'https://ex.io', tls_verify: false)
    restup = Thoom::RestUp.new(config)

    restup.method = 'get'
    restup.endpoint = '/api'

    http = restup.http
    assert_equal true, http.use_ssl?
    assert_equal OpenSSL::SSL::VERIFY_NONE, http.verify_mode
  end

  def test_body_empty
    config = Thoom::HashConfig.new(url: 'http://ex.io')
    restup = Thoom::RestUp.new(config)

    restup.method = 'post'
    restup.endpoint = '/api'

    # If nothing is passed, content-length should be zero
    request = restup.request
    assert_equal 0, request.each_header.to_h['content-length'].to_i
  end

  def test_body
    config = Thoom::HashConfig.new(url: 'http://ex.io')
    restup = Thoom::RestUp.new(config)

    body = 'This is a test body'

    restup.method = 'post'
    restup.endpoint = '/api'
    restup.data = body

    # Make sure the lengths match
    request = restup.request
    assert_equal body, request.body
    assert_equal body.length, request.each_header.to_h['content-length'].to_i
  end

  def test_headers
    config = Thoom::HashConfig.new(
      url: 'http://ex.io', headers: { 'X-Random' => 'abc123' }
    )

    restup = Thoom::RestUp.new(config)
    assert_equal 'abc123', restup.headers[:'X-Random']

    # Make sure that when merging, the new overwrites the old
    restup.headers = { 'X-Random' => 'def456', 'X-Random2' => 'ghi789' }
    assert_equal 'def456', restup.headers[:'X-Random']

    # New headers are added to the hash
    assert_equal 'ghi789', restup.headers[:'X-Random2']
  end

  def test_request_headers
    config = Thoom::HashConfig.new(
      url: 'http://ex.io', headers: { 'X-Random' => 'abc123' }
    )

    restup = Thoom::RestUp.new(config)
    restup.method = 'get'
    restup.endpoint = 'http://ex.io/api'

    # Headers added to the request
    request = restup.request
    assert_equal 'abc123', request.each_header.to_h['x-random']
  end
end


test/output_builder_test.rb
require 'byebug' if ENV['DEBUG']
require 'minitest/autorun'
require_relative '../lib/output_builder'

# Only a simple test of the output builder class
class TestOutputBuilder < Minitest::Test
  def test_simple
    output = Thoom::SimpleOutputBuilder.new
    assert_equal :default, output.colors[:title_color]
  end
  
  def test_default
    output = Thoom::DefaultOutputBuilder.new
    assert_equal 16, output.colors.length
  end
end


lib/rest_up.rb
require 'yaml'
require 'logger'
require 'net/http'
require 'openssl'
require 'uri'
require 'config'
require 'constants'

module Thoom
  # General Error message returned by the class
  class RestUpError < RuntimeError
    attr_reader :message

    def initialize(message)
      @message = message
    end
  end

  # Makes the request
  class RestUp
    attr_accessor :endpoint, :data, :cert
    attr_reader :headers, :log, :method

    def initialize(config = nil)
      @config = config.nil? ? HashConfig.new : config
      @log = Logger.new STDOUT

      @uri = nil
      @xmethods = nil
      @headers = @config.get(:headers, {})
      @standard_methods = %w(delete get head options patch post put)
    end

    def headers=(headers)
      headers.each { |key, val| @headers[key.to_sym] = val }
    end

    def method=(method)
      method.downcase!

      unless @standard_methods.include?(method) || xmethods.include?(method)
        raise RestUpError, 'Invalid Method'
      end

      if xmethods.include? method
        headers['x-http-method-override'] = method.upcase
        method = 'post'
      end

      @method = method
    end

    def request
      raise RestUpError, 'Invalid URL' unless uri.respond_to?(:request_uri)

      request = create_request(uri.request_uri)

      add_request_headers(request)
      add_request_body(request)

      request
    end

    def http
      http = Net::HTTP.new(uri.host, uri.port)
      http.read_timeout = @config.get(:timeout, 300)

      configure_tls http
      configure_client_cert http

      http
    end

    def submit(request)
      http.request request
    end

    def uri
      return @uri if @uri
      @uri = URI.parse url
    end

    def url
      return endpoint if endpoint.start_with?('http')

      @config.get(:url, '') + endpoint
    end

    private

    def create_request(request_uri)
      request = Net::HTTP.const_get(method.capitalize).new request_uri

      configure_basic_auth(request)

      request['User-Agent']  = 'Thoom::RestUp/' + Constants::VERSION
      request.content_length = 0

      request
    end

    def configure_basic_auth(request)
      user = @config.get(:user, '')
      pass = @config.get(:pass, '')

      request.basic_auth(user, pass) unless user.to_s.empty? || pass.to_s.empty?
    end

    def configure_client_cert(http)
      pem = cert.nil? ? @config.get(:cert, '') : cert
      return if pem.empty?

      begin
        http.cert = OpenSSL::X509::Certificate.new pem
        http.key  = OpenSSL::PKey::RSA.new pem
      rescue OpenSSL::OpenSSLError
        raise RestUpError, 'Invalid client certificate'
      end
    end

    def configure_tls(http)
      return if uri.scheme != 'https'
      http.use_ssl = true

      mode = @config.get(:tls_verify, true) ? 'VERIFY_PEER' : 'VERIFY_NONE'
      http.verify_mode = OpenSSL::SSL.const_get(mode)
    end

    def add_request_body(request)
      return if data.nil? || data.empty?

      body = data.clone
      request.content_length = body.length
      request.body = body
    end

    def add_request_headers(request)
      return unless headers.respond_to? :each

      headers.each { |key, val| request[key.to_s.strip] = val.strip }
    end

    def xmethods
      return @xmethods if @xmethods

      xmethods = @config.get(:xmethods, [])
      unless xmethods.respond_to? :map
        raise RestUpError, 'Invalid xmethods configuration'
      end

      @xmethods = xmethods.map(&:downcase)
    end
  end
end


lib/output_builder.rb
require 'constants'

require 'json'
require 'paint'
require 'rexml/document'
require 'rexml/formatters/pretty'

module Thoom
  class OutputBuilder
    attr_accessor :colors, :title_output, :response_time

    def initialize(colors)
      @colors = colors
      @output = ''
    end

    def title(centered = true)
      return if title_output

      client_copy = "Thoom::RestUp v#{Thoom::Constants::VERSION}"
      author_copy = '@author Z.d. Peacock <zdp@thoomtech.com>'
      link_copy   = '@link http://github.com/thoom/restup'

      if centered
        max         = [client_copy.length, author_copy.length, link_copy.length].max + 2
        client_copy = client_copy.center(max, ' ')
        author_copy = author_copy.center(max, ' ')
        link_copy   = link_copy.center(max, ' ')
      end

      @title_output = true
      puts "\n",
           Paint[client_copy, colors[:title_color], colors[:title_bgcolor]],
           Paint[author_copy, colors[:subtitle_color], colors[:subtitle_bgcolor]],
           Paint[link_copy, colors[:subtitle_color], colors[:subtitle_bgcolor]]
    end

    def header(h)
      len = Paint.unpaint(h).length
      l   = '-' * len
      puts "\n#{h}\n#{l}\n"
    end

    def help(config_file, opts)
      title
      section 'How to use RestUp'

      puts <<TEXT
RestUp works out of the box with APIs that use Basic Authentication (though this is not required).
To use other forms of authentication, custom headers can either be passed with each request
or stored in the config file as described below.

If a #{Paint[config_file, colors[:help_filename]]} file exists, the client will pull in defaults and provide several shortcut methods
that can simplify using a REST-based API.

If the API uses form encoded input, you can define your post in JSON format. The client
will encode it automatically.
TEXT

      section 'Console'
      puts opts

      section 'YAML config'
      puts <<TEXT
If a file is not passed in with the `-c, --config` flag, then it will use the default #{Paint[config_file, colors[:help_filename]]}.
The client uses two different methods to find the YAML configuration file. It will
first look in the current directory. If it is not present, it will then look in the current user's
home directory.

This makes it possible to use restup to connect to different APIs simply by changing folders.

KEY          DESC
----         -----
user:       The username. Default: blank, so disable Basic Authentication
pass:       The password. Default: blank, so disable Basic Authentication

url:        The base REST url
json:       The default JSON MIME type. Default: "application/json"
xml:        The default XML MIME type.  Default: "application/xml"

colors:     Hash of default color values
  success:  Color to highlight successful messages.  Default: :green
  warning:  Color to highlight warning messages.     Default: :yellow
  info:     Color to highlight info messages.        Default: :yellow
  error:    Color to highlight error messages.       Default :red

flags:      Default command line options
  display:  What to display by default.
            Values: concise, response_only, response_code_only, succcess_only, verbose
            Default: response_only

headers:    Hash of default headers. Useful for custom headers or headers used in every request.
timeout:    The number of seconds to wait for a response before timing out. Default: 300
tls_verify: Whether or not to verify TLS (SSL) certificates.  Values: true, false.  Default: true

xmethods:   Array of nonstandard methods that are accepted by the API. To use these methods the
            API must support X-HTTP-Method-Override.
TEXT

      section 'Examples'

      header 'GET Request'

      puts <<TEXT
The YAML config:
url: http://example.com/api
user: myname
pass: P@ssWord

#{Paint['restup -j /hello/world', colors[:help_sample_request]]}

To use without the config:
#{Paint['restup -u myname -p P@ssWord -j http://example.com/api/hello/world', colors[:help_sample_request]]}

Submits a GET request to #{Paint['http://example/api/hello/world', colors[:help_sample_url]]} with Basic Auth header using the
user and pass values in the config.

It would return JSON values. If successful, the JSON would be parsed and highlighted in #{Paint[colors[:success].to_s.upcase, colors[:success]]}. If
the an error was returned (an HTTP response code >= 400), the body would be in #{Paint[colors[:error].to_s.upcase, colors[:error]]}.
TEXT

      header 'POST Request'

      puts <<TEXT
The YAML config:
url: http://example.com/api
user: myname
pass: P@ssWord
headers:
  X-Custom-Id: abc12345

#{Paint['restup -m post -j /hello/world < salutation.json', colors[:help_sample_request]]}

Submits a POST request to #{Paint['http://example/api/hello/world', colors[:help_sample_url]]} with Basic Auth header
using the user and pass values in the config. It imports the salutation.json and passes it to the API as application/json
content type. It would also set the X-Custom-Id header with every request.

It would return JSON values. If successful, the JSON would be parsed and highlighted in #{Paint[colors[:success].to_s.upcase, colors[:success]]}. If
the an error was returned (an HTTP response code >= 400), the body would be in #{Paint[colors[:error].to_s.upcase, colors[:error]]}.
TEXT
      exit
    end

    def section(h)
      len = Paint.unpaint(h).length
      l   = '-' * (len + 4)
      puts "\n#{l}\n| #{h} |\n#{l}\n"
    end

    def xp(xml_text)
      out = ''

      formatter = REXML::Formatters::Pretty.new
      formatter.compact = true
      formatter.write(REXML::Document.new(xml_text), out)
      out
    end

    def request(client, request, verbose)
      path  = client.uri.path
      query = ''
      query += '?' + client.uri.query if client.uri.query

      port_color      = client.uri.port == 80 ? :request_port_http : :request_port_tls
      request_section = "REQUEST: #{Paint[client.method.upcase, colors[:request_method]]} "
      request_section += Paint["#{client.uri.host}:", colors[:request_path]]
      request_section += Paint[client.uri.port.to_s, colors[port_color]]
      request_section += Paint[path, colors[:request_path]]
      request_section += Paint[query, colors[:request_endpoint]]

      unless request.respond_to? 'each_header'
        header 'MALFORMED REQUEST'
        quit Paint[request, colors[:error]]
      end

      if verbose
        section request_section if verbose

        header 'HEADERS'
        request.each_header { |k, v| puts "#{k}: #{v}\n" }

        if client.data
          header 'BODY'

          begin
            puts %w(UTF-8 ASCII-8BIT).include?(client.data.encoding.to_s) ? client.data : client.data.encode('ASCII-8BIT')
          rescue EncodingError
            puts "Data posted, but contains non-UTF-8 data, so it's not echoed here."
          end
        end
      else
        puts "\n#{request_section}"
      end
    end

    def response(response, verbose)
      response_color   = response.code.to_i < 400 ? colors[:success] : colors[:error]
      response_section = "RESPONSE: #{Paint[response.code, response_color]} (#{response_time} sec)"

      if verbose || response_color == colors[:error]
        section response_section
        header 'HEADERS'
        response.each_header { |k, v| puts "#{k}: #{v}\n" }
      else
        puts response_section
      end

      header 'BODY' if verbose
      puts 'BODY:' unless verbose

      if !response.body || response.body.empty?
        puts Paint['NONE', colors[:info]]
      else
        body = response.body
        begin
          body.encode!('ASCII-8BIT') if body.encoding.to_s != 'ASCII-8BIT'

          body = if response['content-type'].nil?
                   body
                 elsif response['content-type'].include? 'json'
                   JSON.pretty_unparse(JSON.parse(body))
                 elsif response['content-type'].include? 'xml'
                   xp(body)
                 else
                   body
             end
          puts Paint[body, response_color]
        rescue EncodingError
          puts Paint["RESPONSE contains non-UTF-8 data, so it's not echoed here.", colors[:info]]
        end
      end
    end

    def save_response(response, content_disposition, output)
      if content_disposition && output.nil? && response.to_hash.key?('content-disposition')
        cd = response['content-disposition']
        output = cd[cd.index('filename=') + 9..-1]
      end

      unless output.nil?
        file = File.expand_path(output)
        if File.exist?(File.dirname(file))
          File.open(file, 'w') { |f| f.write response.body }
          puts Paint["Response written to file: #{file}", colors[:info]]
        else
          puts Paint["Could not write to file #{file}", colors[:error]]
        end
      end
    end

    def quit(content, centered = true)
      title(centered)
      puts "\n#{content}"
      exit
    end
  end

  # Sets up the default color set
  class DefaultOutputBuilder < OutputBuilder
    def initialize
      colors = {
        title_color: '4D7326',
        title_bgcolor: :white,

        subtitle_color: :white,
        subtitle_bgcolor: '4D7326',

        help_filename: :yellow,
        help_sample_request: :magenta,
        help_sample_url: :blue,

        request_method: :cyan,
        request_path: '813b5e',
        request_port_http: '813b5e',
        request_port_tls: '264d73',
        request_endpoint: :yellow,

        success: '277326',
        warning: :yellow,
        info: :yellow,
        error: 'c20f12'
      }
      super(colors)
    end
  end

  # Outputs just the basic default colors
  class SimpleOutputBuilder < OutputBuilder
    def initialize
      colors = {
        subtitle_color: :default,
        subtitle_bgcolor: :default,
        title_color: :default,
        title_bgcolor: :default
      }
      super(colors)
    end
  end
end
